web/packages/teleterm/src/services/appUpdater/appUpdater.ts

```diff
@@ -89,15 +107,202 @@ export class AppUpdater {
+          }
+        }
+      }
+    });
   }
 
   dispose(): void {
     this.unregisterEventHandlers();
   }
 
+  /**
+   * Checks for app updates.
+   *
+   * This method enhances the standard autoUpdater.checkForUpdates() by adding
+   * the following behaviors:
+   * - It allows update checks during an ongoing download process.
+   * If a new update is found (or no update is available), the current download
+   * is canceled.
+   * - If an update is checked after a download has completed, the updater
+   * automatically transitions to the `update-downloaded` state.
+   */
+  async checkForUpdates(): Promise<void> {
+    if (this.checkForUpdatesPromise) {
+      this.logger.info('Check for updates already in progress.');
+      return this.checkForUpdatesPromise;
+    }
+
+    this.checkForUpdatesPromise = this.doCheckForUpdates();
+    try {
+      await this.checkForUpdatesPromise;
+    } catch (error) {
+      this.logger.error('Failed to check for updates.', error);
+    } finally {
+      this.checkForUpdatesPromise = undefined;
+    }
+  }
+
+  /** Not safe for concurrent use. */
+  private async doCheckForUpdates(
+    opts: {
+      /**
+       * Whether this is a retry attempt.
+       * Used as a guard to prevent infinite loops.
+       */
+      hasRetried?: boolean;
+    } = {}
+  ): Promise<void> {
+    const result = await autoUpdater.checkForUpdates();
+
+    const newSha = result.updateInfo?.files[0]?.sha512;
+    const oldSha = this.updateCheckResult?.updateInfo.files[0]?.sha512;
+    const isSameUpdate = newSha && oldSha && newSha === oldSha;
+
+    this.updateCheckResult = result;
+
+    const updateUnavailable = !result.isUpdateAvailable;
+    const updateChanged = !isSameUpdate;
+
+    let downloadCanceled = false;
+    if (updateUnavailable || updateChanged) {
+      downloadCanceled = await this.cancelDownload();
+    }
+
+    if (
+      this.shouldAutoDownload() ||
+      // This can occur if the user manually downloads an update
+      // and then triggers another check for updates.
+      // Since the update is already downloaded, the updater should transition
+      // to the `update-downloaded` state automatically.
+      // The update file will be read from the local cache.
+      this.isUpdateDownloaded
+    ) {
+      void this.download();
+      return;
+    }
+
+    // Retry to refresh the state so that the UI won't be showing
+    // a cancellation error.
+    if (downloadCanceled && !opts.hasRetried) {
+      await this.doCheckForUpdates({ hasRetried: true });
+    }
+  }
+
+  /** Starts download. */
+  async download(): Promise<string[]> {
+    if (this.downloadPromise) {
+      this.logger.info('Download already in progress.');
+      return this.downloadPromise;
+    }
+
+    this.downloadPromise = autoUpdater.downloadUpdate();
+    try {
+      await this.downloadPromise;
+      this.isUpdateDownloaded = true;
+    } catch (error) {
+      this.logger.error('Failed to download update.', error);
```

Comment:
Does this error end up being surfaced to the user? I assume this might happen through `autoUpdater`. It'd be good to leave a comment about how this error gets surfaced.

Reply:
Errors that happen in `autoUpdater.checkForUpdates` or `autoUpdater.download` are emitted as events. I added comments.